File: sdk/python/cap/provider.py
"""CAP Provider SDK - for agents that provide services to other agents."""
import hashlib
import json
import time
import threading
import urllib.request
import urllib.error
from typing import Callable, Optional
from cap.models import AgentProfile, SLATerms


class CAPProvider:
    """Provider SDK for agents that offer services on CAP."""

    def __init__(self, agent_id: str, cap_url: str = "http://localhost:4000"):
        self.agent_id = agent_id
        self.cap_url = cap_url.rstrip("/")
        self._handlers: dict[str, Callable] = {}
        self._running = False
        self._processed_requests: set[str] = set()
        self._processed_escrows: set[str] = set()

    def _request(self, method: str, path: str, data: dict = None) -> dict:
        url = f"{self.cap_url}{path}"
        body = json.dumps(data).encode() if data else None
        req = urllib.request.Request(
            url,
            data=body,
            method=method,
            headers={
                "Content-Type": "application/json",
                "X-CAP-Agent-Id": self.agent_id,
            },
        )
        try:
            with urllib.request.urlopen(req) as resp:
                return json.loads(resp.read().decode())
        except urllib.error.HTTPError as e:
            error_body = e.read().decode() if e.fp else ""
            raise RuntimeError(f"CAP API error {e.code}: {error_body}")

    # --- Registration ---

    def register(self, name: str, description: str = "", capabilities: list = None, endpoint: str = "") -> AgentProfile:
        result = self._request("POST", "/cap/v1/agents", {
            "agentId": self.agent_id,
            "name": name,
            "description": description,
            "capabilities": capabilities or [],
            "endpoint": endpoint,
        })
        return AgentProfile(
            agent_id=self.agent_id,
            name=name,
            description=description,
            capabilities=capabilities or [],
            endpoint=endpoint,
            contract_id=result.get("contractId"),
        )

    def create_listing(self, capability: str, description: str = "", base_price: float = 0.0) -> dict:
        return self._request("POST", "/cap/v1/services/listings", {
            "agentId": self.agent_id,
            "capability": capability,
            "description": description,
            "basePrice": base_price,
        })

    # --- Service Handler Registration ---

    def on_request(self, capability: str):
        """Decorator to register a handler for service requests.

        Usage:
            @provider.on_request("translation")
            def handle_translation(params: dict) -> dict:
                return {"translated": translate(params["text"])}
        """
        def decorator(func: Callable):
            self._handlers[capability] = func
            return func
        return decorator

    # --- Offer & Deliver ---

    def make_offer(self, request_contract_id: str, price: float, estimated_time_ms: int = 5000) -> dict:
        return self._request("POST", f"/cap/v1/services/requests/{request_contract_id}/offer", {
            "provider": self.agent_id,
            "price": price,
            "estimatedTimeMs": estimated_time_ms,
        })

    def deliver(self, escrow_contract_id: str, result_hash: str, result_url: str = "") -> dict:
        return self._request("POST", f"/cap/v1/escrows/{escrow_contract_id}/deliver", {
            "provider": self.agent_id,
            "resultHash": result_hash,
            "resultUrl": result_url,
        })

    # --- Polling Loop ---

    def _poll_loop(self, poll_interval: float = 2.0):
        """Poll for incoming service requests and handle them."""
        while self._running:
            try:
                result = self._request("GET", f"/cap/v1/services/requests?as={self.agent_id}")
                requests = result.get("requests", [])

                for req in requests:
                    contract_id = req["contractId"]
                    if contract_id in self._processed_requests:
                        continue

                    capability = req.get("capability", "")
                    handler = self._handlers.get(capability)
                    if not handler:
                        continue

                    self._processed_requests.add(contract_id)

                    params = req.get("params", "{}")
                    if isinstance(params, str):
                        try:
                            params = json.loads(params)
                        except json.JSONDecodeError:
                            params = {"raw": params}

                    max_price = float(req.get("maxPrice", "1.0"))

                    # Make offer
                    print(f"  [{self.agent_id}] Received request for '{capability}', offering...")
                    try:
                        self.make_offer(contract_id, price=max_price * 0.5, estimated_time_ms=3000)
                    except Exception as e:
                        print(f"  [{self.agent_id}] Offer error: {e}")
                        continue

                # Check for escrows that need delivery
                escrows = self._request("GET", f"/cap/v1/escrows?as={self.agent_id}")
                active = escrows.get("escrows", [])

                for escrow in active:
                    escrow_id = escrow["contractId"]
                    if escrow_id in self._processed_escrows:
                        continue

                    capability = escrow.get("capability", "")
                    handler = self._handlers.get(capability)
                    if not handler:
                        continue

                    self._processed_escrows.add(escrow_id)

                    params = escrow.get("params", "{}")
                    if isinstance(params, str):
                        try:
                            params = json.loads(params)
                        except json.JSONDecodeError:
                            params = {"raw": params}

                    print(f"  [{self.agent_id}] Executing '{capability}'...")
                    try:
                        output = handler(params)
                        result_hash = "sha256:" + hashlib.sha256(json.dumps(output, sort_keys=True, default=str).encode()).hexdigest()
                        result_url = f"cap://results/{self.agent_id}/{result_hash}"

                        self.deliver(escrow_id, result_hash, result_url)
                        print(f"  [{self.agent_id}] Delivered result for '{capability}'")
                    except Exception as e:
                        print(f"  [{self.agent_id}] Handler error: {e}")

            except Exception as e:
                if self._running:
                    print(f"  [{self.agent_id}] Poll error: {e}")

            time.sleep(poll_interval)

    def serve(self, poll_interval: float = 2.0, background: bool = False):
        """Start serving requests.

        Args:
            poll_interval: Seconds between polling for new requests.
            background: If True, runs in a background thread.
        """
        self._running = True
        print(f"  [{self.agent_id}] Serving: {list(self._handlers.keys())}")

        if background:
            thread = threading.Thread(target=self._poll_loop, args=(poll_interval,), daemon=True)
            thread.start()
            return thread
        else:
            try:
                self._poll_loop(poll_interval)
            except KeyboardInterrupt:
                self.stop()

    def stop(self):
        """Stop serving."""
        self._running = False
        print(f"  [{self.agent_id}] Stopped.")

File: sdk/typescript/src/provider.ts
/**
 * CAP Provider SDK -- for agents that provide services to other agents.
 */

import { createHash } from "node:crypto";

import type {
  AgentProfile,
  EscrowsResponse,
  RegisterResponse,
  RequestsResponse,
  ServeOptions,
} from "./models.js";

/**
 * A handler function that receives request params and returns a result object.
 */
export type RequestHandler = (
  params: Record<string, unknown>,
) => Record<string, unknown> | Promise<Record<string, unknown>>;

export class CAPProvider {
  readonly agentId: string;
  readonly capUrl: string;

  private _handlers: Map<string, RequestHandler> = new Map();
  private _running = false;
  private _processedRequests: Set<string> = new Set();
  private _processedEscrows: Set<string> = new Set();
  private _pollTimer: NodeJS.Timeout | null = null;

  constructor(agentId: string, capUrl: string = "http://localhost:4000") {
    this.agentId = agentId;
    this.capUrl = capUrl.replace(/\/+$/, "");
  }

  // -------------------------------------------------------------------------
  // Internal HTTP helper
  // -------------------------------------------------------------------------

  private async _request<T = Record<string, unknown>>(
    method: string,
    path: string,
    data?: Record<string, unknown>,
  ): Promise<T> {
    const url = `${this.capUrl}${path}`;
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      "X-CAP-Agent-Id": this.agentId,
    };
    const init: { method: string; headers: Record<string, string>; body?: string } = {
      method,
      headers,
    };
    if (data !== undefined) {
      init.body = JSON.stringify(data);
    }

    const resp = await fetch(url, init);
    const body = await resp.text();

    if (!resp.ok) {
      throw new Error(`CAP API error ${resp.status}: ${body}`);
    }

    return body ? (JSON.parse(body) as T) : ({} as T);
  }

  // -------------------------------------------------------------------------
  // Registration
  // -------------------------------------------------------------------------

  async register(
    name: string,
    description: string = "",
    capabilities: string[] = [],
    endpoint: string = "",
  ): Promise<AgentProfile> {
    const result = await this._request<RegisterResponse>("POST", "/cap/v1/agents", {
      agentId: this.agentId,
      name,
      description,
      capabilities,
      endpoint,
    });
    return {
      agentId: this.agentId,
      name,
      description,
      capabilities,
      endpoint,
      contractId: result.contractId,
    };
  }

  async createListing(
    capability: string,
    description: string = "",
    basePrice: number = 0.0,
  ): Promise<Record<string, unknown>> {
    return this._request("POST", "/cap/v1/services/listings", {
      agentId: this.agentId,
      capability,
      description,
      basePrice,
    });
  }

  // -------------------------------------------------------------------------
  // Service Handler Registration
  // -------------------------------------------------------------------------

  /**
   * Register a handler for incoming service requests of the given capability.
   *
   * This is the TypeScript equivalent of the Python decorator pattern. Usage:
   *
   * ```ts
   * provider.onRequest("translation", async (params) => {
   *   return { translated: translate(params.text as string) };
   * });
   * ```
   *
   * Returns the provider instance for chaining.
   */
  onRequest(capability: string, handler: RequestHandler): this {
    this._handlers.set(capability, handler);
    return this;
  }

  /**
   * Decorator-style handler registration (for use with experimental
   * decorators or as a higher-order function).
   *
   * ```ts
   * const handle = provider.handler("translation");
   * // `handle` is a decorator: (target) => target
   * ```
   */
  handler(capability: string): (fn: RequestHandler) => RequestHandler {
    return (fn: RequestHandler) => {
      this._handlers.set(capability, fn);
      return fn;
    };
  }

  // -------------------------------------------------------------------------
  // Offer & Deliver
  // -------------------------------------------------------------------------

  async makeOffer(
    requestContractId: string,
    price: number,
    estimatedTimeMs: number = 5000,
  ): Promise<Record<string, unknown>> {
    return this._request(
      "POST",
      `/cap/v1/services/requests/${requestContractId}/offer`,
      {
        provider: this.agentId,
        price,
        estimatedTimeMs,
      },
    );
  }

  async deliver(
    escrowContractId: string,
    resultHash: string,
    resultUrl: string = "",
  ): Promise<Record<string, unknown>> {
    return this._request("POST", `/cap/v1/escrows/${escrowContractId}/deliver`, {
      provider: this.agentId,
      resultHash,
      resultUrl,
    });
  }

  // -------------------------------------------------------------------------
  // Polling Loop
  // -------------------------------------------------------------------------

  private async _pollOnce(): Promise<void> {
    try {
      // --- Handle incoming service requests ---
      const reqResult = await this._request<RequestsResponse>(
        "GET",
        `/cap/v1/services/requests?as=${encodeURIComponent(this.agentId)}`,
      );
      const requests = reqResult.requests ?? [];

      for (const req of requests) {
        const contractId = req.contractId;
        if (this._processedRequests.has(contractId)) {
          continue;
        }

        const capability = req.capability ?? "";
        const handlerFn = this._handlers.get(capability);
        if (!handlerFn) {
          continue;
        }

        this._processedRequests.add(contractId);

        const maxPrice = Number(req.maxPrice ?? 1.0);

        console.log(`  [${this.agentId}] Received request for '${capability}', offering...`);
        try {
          await this.makeOffer(contractId, maxPrice * 0.5, 3000);
        } catch (err) {
          console.error(`  [${this.agentId}] Offer error:`, err);
          continue;
        }
      }

      // --- Handle escrows that need delivery ---
      const escrowResult = await this._request<EscrowsResponse>(
        "GET",
        `/cap/v1/escrows?as=${encodeURIComponent(this.agentId)}`,
      );
      const escrows = escrowResult.escrows ?? [];

      for (const escrow of escrows) {
        const escrowRecord = escrow as Record<string, unknown>;
        const escrowId = escrowRecord.contractId as string;
        if (this._processedEscrows.has(escrowId)) {
          continue;
        }

        const capability = (escrowRecord.capability ?? "") as string;
        const handlerFn = this._handlers.get(capability);
        if (!handlerFn) {
          continue;
        }

        this._processedEscrows.add(escrowId);

        let params = escrowRecord.params;
        if (typeof params === "string") {
          try {
            params = JSON.parse(params);
          } catch {
            params = { raw: params };
          }
        }
        if (params === undefined || params === null) {
          params = {};
        }

        console.log(`  [${this.agentId}] Executing '${capability}'...`);
        try {
          const output = await Promise.resolve(
            handlerFn(params as Record<string, unknown>),
          );
          const serialized = JSON.stringify(output, Object.keys(output).sort());
          const resultHash = "sha256:" + this._sha256(serialized);
          const resultUrl = `cap://results/${this.agentId}/${resultHash}`;

          await this.deliver(escrowId, resultHash, resultUrl);
          console.log(`  [${this.agentId}] Delivered result for '${capability}'`);
        } catch (err) {
          console.error(`  [${this.agentId}] Handler error:`, err);
        }
      }
    } catch (err) {
      if (this._running) {
        console.error(`  [${this.agentId}] Poll error:`, err);
      }
    }
  }

  private async _pollLoop(pollInterval: number): Promise<void> {
    while (this._running) {
      await this._pollOnce();
      await this._sleep(pollInterval * 1000);
    }
  }

  // -------------------------------------------------------------------------
  // Serve
  // -------------------------------------------------------------------------

  /**
   * Start serving requests.
   *
   * @param options.pollInterval - Seconds between polls (default 2).
   * @param options.background   - If true, poll in the background and return
   *                               an AbortController you can use to stop.
   */
  serve(options: ServeOptions = {}): AbortController | Promise<void> {
    const { pollInterval = 2.0, background = false } = options;

    this._running = true;
    const capabilities = Array.from(this._handlers.keys());
    console.log(`  [${this.agentId}] Serving: ${JSON.stringify(capabilities)}`);

    if (background) {
      const controller = new AbortController();

      // Start polling in the background -- we intentionally do not await.
      const loop = (async () => {
        try {
          await this._pollLoop(pollInterval);
        } catch {
          // stopped
        }
      })();

      // When the caller aborts, stop the provider.
      controller.signal.addEventListener("abort", () => {
        this.stop();
      });

      // Keep a reference so callers can await if they want.
      (controller as unknown as Record<string, unknown>).__loop = loop;

      return controller;
    }

    // Foreground: return the promise (run until stop() is called).
    return this._pollLoop(pollInterval);
  }

  /** Stop serving. */
  stop(): void {
    this._running = false;
    if (this._pollTimer) {
      clearTimeout(this._pollTimer);
      this._pollTimer = null;
    }
    console.log(`  [${this.agentId}] Stopped.`);
  }

  /** Whether the provider is currently running its poll loop. */
  get running(): boolean {
    return this._running;
  }

  // -------------------------------------------------------------------------
  // Helpers
  // -------------------------------------------------------------------------

  /** Compute a hex-encoded SHA-256 hash of the given string. */
  private _sha256(input: string): string {
    return createHash("sha256").update(input).digest("hex");
  }

  private _sleep(ms: number): Promise<void> {
    return new Promise((resolve) => {
      this._pollTimer = setTimeout(resolve, ms);
    });
  }
}
